components/raftstore/src/store/fsm/peer.rs

```diff
@@ -4683,17 +4701,72 @@ where
                 "epoch" => ?region_epoch,
                 "current_epoch" => ?region.get_region_epoch(),
             );
+            // test purpose
+            test_only_callback(cb, self.fsm.peer.region_buckets.clone());
             return;
         }
-        let mut region_buckets = Buckets::default();
-        region_buckets.version = timespec_to_ns(monotonic_raw_now());
-        region_buckets.set_keys(bucket_keys.into());
-        // add region's start/end key
-        region_buckets
-            .keys
-            .insert(0, region.get_start_key().to_vec());
-        region_buckets.keys.push(region.get_end_key().to_vec());
-        self.fsm.peer.region_buckets = region_buckets;
+
+        if let Some(bucket_ranges) = bucket_ranges {
+            assert_eq!(buckets.len(), bucket_ranges.len());
+            let mut region_buckets_keys = self.fsm.peer.region_buckets.keys.to_vec();
+            let mut region_buckets = Buckets::default();
+            region_buckets.version = timespec_to_ns(monotonic_raw_now());
+            let mut i = 0;
+            let mut j = 0;
+            let buckets = buckets.drain(..);
+            for mut bucket in buckets {
+                while i < region_buckets_keys.len() {
+                    if region_buckets_keys[i] == bucket_ranges[j].0 {
+                        // the bucket size is small and does not have split keys,
+                        // then it should be merged with its left neighbor
+                        if bucket.keys.is_empty()
+                            && bucket.size
+                                <= self.ctx.coprocessor_host.cfg.region_bucket_merge_size.0
+                        {
+                            // i is not the last entry (which is end key)
+                            if region_buckets_keys.len() > 2 {
+                                region_buckets_keys.remove(i); // bucket is too small
+                                // the key is removed, to compensate the += 1 below.
+                                i -= 1;
```

Comment:
Why not just `continue`?

Reply:
Good catch. The previous version has some other code to run below, but this version can use 'continue'.